plugins/modules/opkg.py

```diff
@@ -97,144 +97,106 @@
-        else:
-            version_str = ""
-
-        if force:
-            rc, out, err = module.run_command([opkg_path, "install", force, package + version_str])
-        else:
-            rc, out, err = module.run_command([opkg_path, "install", package + version_str])
-
-        if not query_package(module, opkg_path, package, version):
-            module.fail_json(msg="failed to install %s%s: %s" % (package, version_str, out))
-
-        install_c += 1
-
-    if install_c > 0:
-        module.exit_json(changed=True, msg="installed %s package(s)" % (install_c))
-
-    module.exit_json(changed=False, msg="package(s) already present")
-
-
-def main():
-    module = AnsibleModule(
+class Opkg(StateModuleHelper):
+    module = dict(
         argument_spec=dict(
             name=dict(aliases=["pkg"], required=True, type="list", elements="str"),
             state=dict(default="present", choices=["present", "installed", "absent", "removed"]),
             force=dict(default="", choices=["", "depends", "maintainer", "reinstall", "overwrite", "downgrade", "space", "postinstall", "remove",
                                             "checksum", "removal-of-dependent-packages"]),
             update_cache=dict(default=False, type='bool'),
-        )
+        ),
     )
 
-    opkg_path = module.get_bin_path('opkg', True, ['/bin'])
+    def __init_module__(self):
+        self.vars.set("install_c", 0, output=False, change=True)
+        self.vars.set("remove_c", 0, output=False, change=True)
 
-    p = module.params
+        state_map = dict(
+            query="list-installed",
+            present="install",
+            installed="install",
+            absent="remove",
+            removed="remove",
+        )
 
-    if p["update_cache"]:
-        update_package_db(module, opkg_path)
+        def _force(value):
+            if value == "":
+                value = None
+            return cmd_runner_fmt.as_optval("--force-")(value, ctx_ignore_none=True)
+
+        self.runner = CmdRunner(
+            self.module,
+            command="opkg",
+            arg_formats=dict(
+                package=cmd_runner_fmt.as_list(),
+                state=cmd_runner_fmt.as_map(state_map),
+                force=cmd_runner_fmt.as_func(_force),
+                update_cache=cmd_runner_fmt.as_bool("update")
+            ),
+        )
 
-    pkgs = p["name"]
+    @staticmethod
+    def split_name_and_version(package):
+        """ Split the name and the version when using the NAME=VERSION syntax """
+        splitted = package.split('=', 1)
+        if len(splitted) == 1:
+            return splitted[0], None
+        else:
+            return splitted[0], splitted[1]
+
+    def _package_in_desired_state(self, name, desired_installed, version=None):
+        dummy, out, dummy = self.runner("state package").run(state="query", package=name)
+
+        has_package = out.startswith(name + " - %s" % "" if not version else (version + " "))
```

Comment:
Without the extra bracket, the code does not work if `version` is actually provided (i.e. not `None`). The function returns `False` even if the packet is installed in the correct version.
Unfortunately this is still not covered by a testcase (otherwise this would have been caught) -> @russoz , could you add one?

Reply:
@joergho Errr... I am having a different problem here. I have an OpenWRT router running here, I tried running `opkg install package=version` and that syntax does not seem to be accepted by `opkg`.

Checking the docs: https://openwrt.org/docs/guide-user/additional-software/opkg
it also does not mention any syntax to specify the version. In the OpenWRT forum, searching very shallowly, whenever some asks "how to install a specific version of X?" it seems the answer is, most of the times: don't do that. If you really must, then we suggest you build your custom image. Or you can download the `.ipk` file and install it directly.

Reply:
Here:
```bash
root@OpenWrt:~# opkg list-installed | grep zlib
root@OpenWrt:~# opkg list | grep ^zlib
zlib - 1.2.11-6 - zlib is a lossless data-compression library. This package includes the shared library.
zlib-dev - 1.2.11-6 - zlib is a lossless data-compression library. This package includes the development support files.
root@OpenWrt:~# opkg install zlib=1.2.11-6
Unknown package 'zlib=1.2.11-6'.
Collected errors:
 * opkg_install_cmd: Cannot install package zlib=1.2.11-6.
root@OpenWrt:~# opkg install zlib=1.2.11
Unknown package 'zlib=1.2.11'.
Collected errors:
 * opkg_install_cmd: Cannot install package zlib=1.2.11.
root@OpenWrt:~# opkg list | grep ^opkg
opkg - 2022-02-24-d038e5b6-1 - Lightweight package management system opkg is the opkg Package Management System, for handling installation and removal of packages on a system. It can recursively follow dependencies and download all packages necessary to install a particular package.  opkg knows how to install both .ipk and .deb packages.
```

Reply:
Hm... Then it seems that the the OpenWRT flavored opkg (mainted at [1]) does not offer the same feature as the Yocto/OpenEmbedded flavored opkg (maintained at [2]). A lot of embedded Linux distributions (including NI Linux RT) are under the hood Yocto/OpenEmbedded based, therefore I see value in also testing opkg from the Yocto world. Let' talk offline about how this could be integrated as a test. I'll approve the PR for now.

[1] https://git.openwrt.org/project/opkg-lede.git
[2] https://git.yoctoproject.org/opkg/

Reply:
> Hm... Then it seems as if the OpenWRT flavored opkg (maintained at [1]) does not offer the same feature as the Yocto/OpenEmbedded flavored opkg (maintained at [2]). A lot of embedded Linux distributions (including NI Linux RT) are under the hood Yocto/OpenEmbedded based, therefore I see value in also testing opkg from the Yocto world. Let' talk offline about how this could be integrated as a test. I'll approve the PR for now.
> 
> [1] https://git.openwrt.org/project/opkg-lede.git [2] https://git.yoctoproject.org/opkg/

In that case, please do open another PR, as bugfix _and_ docs, to make it explicit that specifying the version is something that is only available on those other systems.

Reply:
And adding a unit testcase for that situation is easy, coming up in few minutes.

Reply:
Except when an edge case is found and one has to spend time debugging it ;-) The output of `opkg list-installed` also seems to be different between those systems. I am one inch away from suggesting you create a new module, say, yocto_opkg or something like that.

Reply:
I extended the docs in #5810. I think as long as the OpenWrt based test is working and the documentation is clear about OpenWrt vs. Yocto based Linux distributions, it is not worth creating a separate module for it.

Reply:
Fair enough. Thanks!